src/main/java/graphql/execution/ExecutionPath.java

```diff
@@ -63,27 +68,48 @@ public ExecutionPath getPathWithoutListEnd() {
         if (ROOT_PATH.equals(this)) {
             return ROOT_PATH;
         }
-        if (segment instanceof StringPathSegment) {
+        if (segment instanceof String) {
             return this;
         }
         return parent;
     }
 
+    /**
+     * @return true if the end of the path has a list style segment eg 'a/b[2]'
+     */
+    public boolean isListSegment() {
+        return segment instanceof Integer;
+    }
+
+    /**
+     * @return true if the end of the path has a named style segment eg 'a/b[2]/c'
+     */
+    public boolean isNamedSegment() {
+        return segment instanceof String;
+    }
+
+
     public String getSegmentName() {
-        if (segment instanceof StringPathSegment) {
-            return ((StringPathSegment) segment).getValue();
-        } else {
-            if (parent == null) {
-                return null;
-            }
-            return ((StringPathSegment) parent.segment).getValue();
-        }
+        return (String) segment;
+    }
+
+    public int getSegmentIndex() {
+        return (int) segment;
+    }
+
+    public Object getSegmentValue() {
+        return segment;
+    }
+
+    public ExecutionPath getParent() {
+        return parent;
     }
```

Comment:
Wow we never had that :)

Reply:
We have drop segment, which is basically the same, right? 
Should we keep drop segment or get parent?

Reply:
both I think

Reply:
I think both are ok in that one is a read style thing - tell me the parent - where was dropSegment feels like a write still thing (but its immuatble so its not)

For backwards compatibility AND understanding - I would have both